# Задачи кавалера де Мере

Решим первую задачу кавалера де Мере: какова вероятность, что при 4 бросках игрального кубика хотя бы раз выпадет 6 очков?

Посчитаем эту вероятность первым способом - зададим вероятностное пространство $\Omega$.

In [2]:
from itertools import product

n = 4
omega = sorted(product(['1', '2', '3', '4', '5', '6'], repeat=n))
print(len(omega))

1296


In [3]:
len(omega) == 6 ** 4

True

Как видим, $|\Omega| = 1296 = 6^4$, поскольку на каждый из 4 бросков есть ровно 6 вариантов результата (число выпавших очков варьируется от 1 до 6).

In [4]:
omega[:15]

[('1', '1', '1', '1'),
 ('1', '1', '1', '2'),
 ('1', '1', '1', '3'),
 ('1', '1', '1', '4'),
 ('1', '1', '1', '5'),
 ('1', '1', '1', '6'),
 ('1', '1', '2', '1'),
 ('1', '1', '2', '2'),
 ('1', '1', '2', '3'),
 ('1', '1', '2', '4'),
 ('1', '1', '2', '5'),
 ('1', '1', '2', '6'),
 ('1', '1', '3', '1'),
 ('1', '1', '3', '2'),
 ('1', '1', '3', '3')]

Определим нужное нам событие  $A$

In [5]:
A = sorted({i for i in omega if i.count('6') >= 1})
A[:15]

[('1', '1', '1', '6'),
 ('1', '1', '2', '6'),
 ('1', '1', '3', '6'),
 ('1', '1', '4', '6'),
 ('1', '1', '5', '6'),
 ('1', '1', '6', '1'),
 ('1', '1', '6', '2'),
 ('1', '1', '6', '3'),
 ('1', '1', '6', '4'),
 ('1', '1', '6', '5'),
 ('1', '1', '6', '6'),
 ('1', '2', '1', '6'),
 ('1', '2', '2', '6'),
 ('1', '2', '3', '6'),
 ('1', '2', '4', '6')]

Определим теперь функцию подсчета вероятности события $X$ при данном вероятностном пространстве.

In [6]:
def prob(X): 
    return len(X)/len(omega)

In [7]:
print(prob(A))

0.5177469135802469


Теперь посчитаем вероятность этого же события другим способом - через вероятность его дополнения $B=A^c$.

In [8]:
B = sorted({i for i in omega if i.count('6') == 0})
B[:15]

[('1', '1', '1', '1'),
 ('1', '1', '1', '2'),
 ('1', '1', '1', '3'),
 ('1', '1', '1', '4'),
 ('1', '1', '1', '5'),
 ('1', '1', '2', '1'),
 ('1', '1', '2', '2'),
 ('1', '1', '2', '3'),
 ('1', '1', '2', '4'),
 ('1', '1', '2', '5'),
 ('1', '1', '3', '1'),
 ('1', '1', '3', '2'),
 ('1', '1', '3', '3'),
 ('1', '1', '3', '4'),
 ('1', '1', '3', '5')]

In [9]:
print(1 - prob(B))

0.5177469135802469


Вероятость события $B$ легко посчитать и вручную: для каждого исхода события $B$ на каждый из его 4 бросков есть ровно 5 вариантов (число выпавших очков варьируется от 1 до 5). Значит, $|B| = 5^4$. Следовательно, $|A| = 6^4 - 5^4$.

In [10]:
p = (6**4 - 5**4) / (6**4)
print(p)

0.5177469135802469


In [11]:
prob(A) == 1 - prob(B) == p

True

Итак, найденная вероятность превосходит 50%, поэтому при большом количестве игр кавалер де Мере будет в выигрыше.

Перейдем ко второй задаче кавалера де Мере: какова вероятность, что при 24 бросках пары игральных кубиков хотя бы раз выпадет 12 очков (то есть выпадут две шестерки)? Посмотрим, сколько исходов возникает при броске пары кубиков.

In [13]:
omega = set(product(['1', '2', '3', '4', '5', '6'], repeat=2))
omega

22452257707354557240087211123792674816

In [14]:
print(len(omega))

36


Проще всего посчитать данную вероятность аналитически, через вероятность дополнения данного события. В каждом исходе из дополнения могут реализоваться все исходы, кроме выпадения двух шестерок, то есть 35 вариантов из 36. Поскольку бросков 24, вероятность дополнения нашего события равна $(\frac{35}{36})^{24}$.

In [19]:
r = 1 - (35 / 36) ** 24
print(r)

0.4914038761309034


Как видим, найденная вероятность меньше 50%, поэтому при большом количестве игр кавалер де Мере будет в проигрыше.